### ISSUE TITLE
Training SynthSeg on spine-generic data

### ISSUE THREAD

**Issue by naga-karthik** [sct-pipeline/contrast-agnostic-softseg-spinalcord#111]
This issue summarizes my attempt at training the [SynthSeg](https://www.sciencedirect.com/science/article/pii/S1361841523000506?via%3Dihub) on the spine-generic multi subject dataset. 

**Brief context:** SynthSeg was originally proposed for the segmentation of brain scans of any resolution and contrast. Because is "contrast-agnostic", I am trying to re-train the model as a baseline for our contrast-agnostic spinal cord (SC) segmentation model. Important note here is that SynthSeg required _fully-labeled_ brain scans from which it synthesizes fake brain images (sampled from a GMM) and then trains the segmentation model. A notable challenge in re-training SynthSeg for SC images is that it would require all the labels in a SC scan (i.e. those of  SC, cerebrospinal fluid, vertebrae, bones, brain, etc.) 

As it is not feasible to obtain labels for each anatomical region in a SC image, I tried to ease the constraints by only focusing on the segmentations of 4 parts: (i) vertebrae, (ii) discs, (iii) SC, and (iv) cerebrospinal fluid (CSF). The labels for these regions were obtained using the TotalSegmentatorMRI model. 

### Experiments

The SynthSeg repo contains well-described tutorials for re-training your own models. The following sections will describe how I have tweaked SynthSeg for SC data.

#### Defining the labels for generating synthetic images

There are 4 key elements in generating synthetic images based on labels:

1. [Defining the path to the training label maps](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/2-generation_explained.py#L38): In my case, this corresponds to the folder containing the TotalSeg labels. I only used the labels for T1w and T2w contrasts as the labels were not available for the rest of the spine-generic contrasts.
2. [Label classes for generation](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/2-generation_explained.py#L49-L61): In my case, these correspond to the ids for the background, vertebrae, discs, SC and CSF. Following the preliminary version of [TotalSpineSeg](https://github.com/neuropoly/totalspineseg) model, I used the following ids to define the labels. Note that using the same ids in the tutorial didn't make sense as those labels correspond to the brain (and not the SC)
```python
gen_labels_cerv_totalseg = np.array([0, 
    # vertebrae (covering C1 to T12)
    41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23,
    # discs (covering C1/C2 to T11/T12)
    224, 223, 222, 221, 220, 219, 218, 217, 216, 215, 214, 213, 212, 211, 210, 209, 208,
    # SC and CSF
    200, 201,
])
```
3. [Defining the output label maps](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/2-generation_explained.py#L71-L81): To keep it simple, I used the same labels as above, meaning that the synthetic images will have _all_ classes that are defined in the label generation array above. (SynthSeg provides an option to disable some labels if you don't want to generate them). Also, to keep it simple, all the vertebrae correspond to one class and all the discs correspond to one class (class id shown below). 
4. [Defining the generation classes](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/2-generation_explained.py#L106-L115): This corresponds to the output classes that the segmentation model learns to output during training. Unlike a random collection of ids as shown above, these are typically incremental ids starting from 0 to the total number of classes. In my case, I provided the following 4 classes: 
```python
1: SC
2: CSF
3: Vertebrae
4: Discs
```

These are essentially the 4 key parameters in the procedure for generating synthetic images from the GMM.

**Comment by naga-karthik** [comment]
The following are a few examples of synthetic images from the GMM 

<details>
<summary> Example 1 </summary>

![ezgif com-animated-gif-maker](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/3b45630e-eccd-4dcc-9552-0233e7b9d59e)


</details>

<details>
<summary> Example 2 </summary>

![ezgif com-animated-gif-maker-2](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/9b691326-4b06-4a20-85d4-cfe330d7ed57)

</details>

<details>
<summary> Example 3 </summary>

![ezgif com-animated-gif-maker-3](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/9643e96c-3260-46b3-a862-ce1c1c6a70bb)

</details>

As expected, the synthetic images look very weird --  which is the point of SynthSeg i.e. the heavy deformation and/or unrealistic images -- which makes is contrast-agnostic.

**Comment by naga-karthik** [comment]
### Training

The training procedure borrows several aspects of the synthetic image generation procedure (described in the 1st comment) and adds the regular training hyperparameters. To keep it close to our contrast-agnostic model for SC, I set the [output shape](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/3-training.py#L60) to `64 x 192 x 256`. I also set `bias_field_std=0.5` as this is the range we chose in [our transforms](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/main/monai/transforms.py#L30). 

When training for 5 epochs initially, the model encountered `NaN` values in the loss. It seems that loss values were also encountered [here](https://github.com/BBillot/SynthSeg/issues/63) when re-training SynthSeg. More debugging to be continued ...

**Comment by naga-karthik** [comment]
I was able to fix the `NaN` issue and the first set of results are in. Each following comment describes the experiment along with the results obtained. The generation labels and the segmentation classes remain the same as described in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/111#issue-2364557862. A total of **209 labeled cervical spine scans** were used consisting of labels for the cord, CSF, vertebrae and intervertebral discs obtained from T1w and T2w contrasts. (Note, SynthSeg primarily used the fully-labeled brain scans for T1w contrast)

#### Experiment 1

The following hyperparameters are defined in [this training script](https://github.com/BBillot/SynthSeg/blob/master/scripts/tutorials/3-training.py), which I have modified to stay close to our contrast-agnostic model for SC segmentation. The hyperparams defined below are the only ones changed for this experiment, rest all are set to their default values. 

```python
n_neutral_labels = None   # (setting this to an incorrect value is what I believe resulted in NaN values above
# shape and resolution of the outputs
target_res = None   # SynthSeg by default outputs everything at 1mm iso; since we don't want that, it is set to None  
output_shape = (64, 192, 256)    # setting this to the original shape of the labels (192, 260, 320) resulted in OOM errors;  
                                 # the cropped output shape is also close to what we use in contrast-agnostic (64, 192, 320) i.e. heavy cropping on R-L dim. 

# spatial deformation parameters
flipping = False     
rotation_bounds = 20  
shearing_bounds = .012
bias_field_std = .5

# architecture parameters
activation = 'relu'     # we used relu, so changed it from 'elu' to 'relu'

# training parameters
dice_epochs = 25       # number of training epochs
steps_per_epoch = 1000  # number of iteration per epoch
batchsize = 2
```

#### Results

<details>
<summary> Test sample 1 </summary>

The prediction is not good as it fails to capture the structure of the cervical spine. None of the labels (i.e. cord, CSF, vertebrae, discs, etc.) are properly segmented. 

**sub-barcelona05_T1w**

<img width="1510" alt="Screenshot 2024-07-03 at 9 51 38 AM" src="https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/e49ae94d-8909-4cd3-afa9-1b9a56ad4428">

</details>


<details>
<summary> Test sample 2 </summary>

**sub-amu05_T1w**

<img width="1510" alt="Screenshot 2024-07-03 at 9 56 18 AM" src="https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/f9959a14-a624-46b7-9cf3-4be5d95e0758">

</details>

We see that the predictions are not good as they fail to capture the structure of the cervical spine. None of the labels (i.e. cord, CSF, vertebrae, discs, etc.) are properly segmented.

**Comment by naga-karthik** [comment]
#### Experiment 2

Since `flipping=True` was used by default in SynthSeg, I trained another model this time setting `flipping` to `True` and keeping rest of the hyperparams defined in https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/111#issuecomment-2206186379 to be the same. 

<details>
<summary> Test sample 1 </summary>

**sub-barcelona05_T1w**

<img width="1510" alt="Screenshot 2024-07-03 at 10 08 41 AM" src="https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/673bbfe4-c316-40ce-8fdb-e4adeb3e5be5">


</details>

Surprisingly, the model does not predict anything when `flipping=True` during training

**Comment by naga-karthik** [comment]
#### Experiment 3

To simplify the segmentation problem, I tried to re-train the model with only 2 output classes (i.e. those of the spinal cord and the CSF). Specifically, given all the following labels, the model is trained to output only the SC and CSF classes with label values 200 and 201 (keeping the rest of the hyperparams same as in [Experiment 1](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/issues/111#issuecomment-2206186379))

```python
gen_labels_cerv_totalseg = np.array([0, 
    # vertebrae (covering C1 to T12)
    41, 40, 39, 38, 37, 36, 35, 34, 33, 32, 31, 30, 29, 28, 27, 26, 25, 24, 23,
    # discs (covering C1/C2 to T11/T12)
    224, 223, 222, 221, 220, 219, 218, 217, 216, 215, 214, 213, 212, 211, 210, 209, 208,
    # SC and CSF --> only output these labels
    200, 201,
])
```

Surprisingly, the model failed to output anything as we get empty predictions for all test images. 

<details>

<summary> Test sample 1</summary>

**sub_brnoUhb04_T1w**

<img width="1510" alt="Screenshot 2024-07-04 at 11 55 28 AM" src="https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/assets/53445351/9ffaa7d8-3045-48fa-9a5d-62ce711b76e4">

</details>

**Comment by jcohenadad** [comment]
> But, to fix this, it would mean that we need all the labels (i.e. for those of the brain and those of the SC, which is really impractical)

indeed-- let's not go there

**Comment by naga-karthik** [comment]
Following an excellent clarification from the reviewer, I realized that I had indeed misunderstood how SynthSeg was generating the synthetic scans. Since the beginning, we have been assuming that SynthSeg requires the labels for _all_ the structures (i.e. those of soft tissues around the cord, the brain i.e. all those _not_ the spinal cord). But that isn't the case. SynthSeg can work even without the GT labels for the "non-interesting structures" (in our case, everything outside the cord segmentation). In the paper, for the cardiac segmentation example, they generated the labels outside the cardiac part by clustering the image based on intensities using an expectation-maximization (EM) algorithm. While EM-based clustering is not a hard requirement, the labels for non-interesting structures can also be created using the (much simpler) [KMeans](https://scikit-learn.org/1.5/modules/generated/sklearn.cluster.KMeans.html) algorithm. What follows is the description of my attempt to create KMeans-based labels for our SC scans.

Following the approach mentioned in Supplement 7 (Figure S3 caption), I kept the GT cord segmentation fixed and generated labels for additional (non-interesting) structures using 3-10 clusters. Script for this can be found [here](https://github.com/sct-pipeline/contrast-agnostic-softseg-spinalcord/blob/nk/r2-synthseg/scripts/generate_kmeans_labels.py). Inputs are T1w scans and their corresponding GT cord masks. Outputs are multi-class labels where 0 is the background, 1 is the cord and 2-11 are the set of labels obtained with Kmeans clustering. Some examples of the synthetic scans are below. 

<details>
<summary> Example 1 </summary>

![Image](https://github.com/user-attachments/assets/943be704-65f8-4f1d-a6d6-7ee5c59d642e)

</details> 


<details>
<summary> Example 2 </summary>

![Image](https://github.com/user-attachments/assets/74adbead-4cc5-409f-939e-2176adb05c6a)


</details>


<details>
<summary> Example 3 </summary>

![Image](https://github.com/user-attachments/assets/f882f02a-7c1a-458b-a854-e39f6dc46f43)

</details>

The synthetic images look similar to the ones show in Supplement 1, Figure S1, confirming that this approach for clustering (for non-brain images) and generation is valid. The next step is train SynthSeg on our data and do a proper comparison with the `contrast-agnostic` model

**Comment by naga-karthik** [comment]
Results from SynthSeg training are in! Thanks to the proper implementation of the clustering, the model did predict something this time! A few examples are shown below:

<details> 

<summary> sub-balgrist06_T1w_synthseg </summary>

![Image](https://github.com/user-attachments/assets/7885612b-7bb4-49f0-87e3-d139df6b8d63)

</details>


<details> 

<summary> sub-balgrist06_T2w_synthseg </summary>

![Image](https://github.com/user-attachments/assets/a9254029-73c7-4a02-966a-f01f69e7a0d4)

</details>

We see that SynthSeg's prediction for T1w is relatively quite good compared to the one for T2w. For T2star and MTon contrasts, the did not predict anything. Note that I only used the T1w images to do the intensity-based clustering following Table 1 in the original paper. Based on these results,  it seems that SynthSeg is "contrast-agnostic" _only as long as the field-of-view of the images are the same._ For T2w, the model did predict parts of the cord only because the field-of-view/resolution is similar to that of T1w. For the rest of the contrasts, the images are highly anistropic and only contain partial views of the cord (in the S-I plane). 

Next step: Running one more experiment where 4 contrasts (i.e. T1w, T2w, T2star, DWI) are used when generating labels using K-Means clustering. Maybe exposure to the labels covering all kinds of fields-of-view in the dataset would improve SynthSeg performance.

**Comment by naga-karthik** [comment]
closing as the purpose of this is resolved.
